### ISSUE TITLE
[mac-v2] Dependency graph — Mac perf-tuning v2 tracking issue

### ISSUE THREAD

**Issue by hiyukoim** [hiyukoim/CoppyLora_webUI#24]
## Goal

Tracking issue for v2 of the macOS Apple Silicon port: performance tuning. v1 (issues #1-#7, all merged) made the app **run** on Mac. v2 makes it **run well**. Target: ≤20 s/iter on M4 Pro 24 GB (down from the v1-observed 85–305 s/iter).

## Scope of v2

- Mac-specific MPS env vars in `start.sh`.
- Python-side memory fraction in `CoppyLora_webUI.py`.
- `sdpa = true` in `config.toml` (Metal-native attention).
- Optional `config_lowmem.toml` (Adafactor + `fused_backward_pass`) for memory-tight machines.
- PyTorch pin `>=2.6` in `install.sh` (avoid 2.5.x regression).
- System hygiene + perf table in `README_MAC.md`.
- Smoke test verifying perf gains.
- (Conditional) layer-#5 patches inside `sd-scripts/` if smoke test still misses target.

**Hard constraint:** Training math and final LoRA quality must be unchanged from Windows / v1. Per-tensor max abs diff ≤ 1e-4 over 50 steps. Same image/seed/caption → visually indistinguishable LoRA outputs.

## Sub-issues

| # | Title | Phase | Blocks |
|---|---|---|---|
| #16 | start.sh — MPS env vars + CPU fallback | perf | #22 |
| #17 | CoppyLora_webUI.py — memory fraction 0.80 | perf | #22 |
| #18 | config.toml — sdpa = true | config | #22 |
| #19 | config_lowmem.toml — Adafactor + fused_backward_pass | config | #22 |
| #20 | install.sh — pin PyTorch >=2.6 | install | #22 |
| #21 | README_MAC.md — system hygiene + perf table | docs | #22 |
| #22 | Smoke test — verify v2 gains, tag v1.1.0-mac | test | #23 |
| #23 | (conditional) Layer-#5 sd-scripts patches | perf | (final) |

## Dependency graph

```
        #16 start.sh
        #17 .py memory fraction
        #18 config.toml sdpa     ─── all parallel ──┐
        #19 config_lowmem.toml                      │
        #20 install.sh pin                          │
        #21 README_MAC.md                           │
                                                    │
                                                    ▼
                                            #22 Smoke test
                                                    │
                                ┌───────────────────┴───────────────────┐
                                ▼                                       ▼
                       Pass: tag v1.1.0-mac           Fail: open #23 (sd-scripts patches)
                                                                       │
                                                                       ▼
                                                              Pass firewall: tag v1.2.0-mac
```

## Implementation order (Phase 3 — atomic, one at a time)

#16 through #21 are **independent of each other** and could be done in parallel sessions. In a single thread:

1. **#18** sdpa flag — smallest change (one line in config.toml), highest expected impact.
2. **#16** start.sh env vars — three exports.
3. **#17** memory fraction in .py — one if-block.
4. **#20** PyTorch pin — one-line change to install.sh.
5. **#19** config_lowmem.toml — new file.
6. **#21** README_MAC.md — append section.
7. **#22** smoke test — verify, tag v1.1.0-mac.
8. **#23** (only if #22 misses target) — sd-scripts patches gated by quality firewall, tag v1.2.0-mac on success.

## Branch & PR strategy

- Working branch: `mac` (continues from v1).
- Each sub-issue gets its own atomic commit.
- One PR per issue into `hiyukoim/CoppyLora_webUI:mac`.
- v2 ships as `v1.1.0-mac` tag. Layer-#5 patches (if needed) ship as `v1.2.0-mac`.

## Decisions locked in (from planning)

- **Quality firewall:** ≤1e-4 per-tensor diff over 50 steps + visual ComfyUI parity check. Applies to any sd-scripts patch (#23). Does **not** apply to the lowmem profile (#19) since switching optimizers is opt-in by design.
- **PyTorch pin:** `>=2.6` (not `==2.11.*`) — gives users access to fixes, blocks the 2.5.x regression.
- **Memory fraction:** `0.80` calibrated for 24 GB M4 Pro. Configurable via env var is a follow-up (post-v2 tag).
- **Default optimizer:** stays AdamW. Adafactor is opt-in via `config_lowmem.toml`.
- **mps-bitsandbytes:** not in v2. Maybe a candidate in #23 if firewall passes (unlikely).

## Done when

- All of #16-#22 closed via merged PRs.
- Smoke test (#22) shows ≤20 s/iter steady-state on M4 Pro and produces a valid `.safetensors`.
- `v1.1.0-mac` release tagged on `hiyukoim/CoppyLora_webUI`.
- (Conditional) #23 closed (either as `not_planned` if #22 passed, or merged with `v1.2.0-mac` tag).

## Reference

- v1 tracking issue: #7
- v1 smoke-test transcript (showing the perf problem we're fixing): `tasks/brslaqkrf.output:158-162`
- Plan file: `/Users/yuko/.claude/plans/i-want-to-create-happy-pebble.md`
